src/audio/tdfb/tdfb_direction.c

```diff
@@ -0,0 +1,497 @@
+		*channel = 0;
+}
+
+static inline int16_t max_mic_distance(struct tdfb_comp_data *cd)
+{
+	int16_t dx;
+	int16_t dy;
+	int16_t dz;
+	int32_t d2;
+	int32_t d2_max = 0;
+	int i;
+	int j;
+
+	/* Max lag based on largest array dimension. Microphone coordinates are Q4.12 meters */
+#ifdef DEBUG
+	for (i = 0; i < cd->config->num_mic_locations; i++) {
+		printf("printf: Mic%d (%d, %d, %d)\n", i, cd->mic_locations[i].x,
+		       cd->mic_locations[i].y, cd->mic_locations[i].z);
+	}
+#endif
+
+	for (i = 0; i < cd->config->num_mic_locations; i++) {
+		for (j = 0; i < cd->config->num_mic_locations; i++) {
+			if (j == i)
+				continue;
+
+			dx = cd->mic_locations[i].x - cd->mic_locations[j].x;
+			dy = cd->mic_locations[i].y - cd->mic_locations[j].y;
+			dz = cd->mic_locations[i].z - cd->mic_locations[j].z;
+
+			/* d2 is Q8.24 meters */
+			d2 = dx * dx + dy * dy + dz * dz;
+			if (d2 > d2_max)
+				d2_max = d2;
+		}
+	}
+
+	/* z = pow(x, y), where x is Q6.26, y is Q2.30, z is Q12.20, the shifts below
+	 * return the max array dimension as meters Q4.12;
+	 */
+	return drc_pow_fixed(d2_max << 2, ZERO_POINT_FIVE) >> 8;
+}
+
+static inline bool line_array_mode_check(struct tdfb_comp_data *cd)
+{
+	int32_t prod;
+	int i;
+	int num_mic_locations = cd->config->num_mic_locations;
+
+	if (num_mic_locations == 2)
+		return true;
+
+	/* TODO expand check to (x,y,z), this is just (x,y) */
+	for (i = 0; i < num_mic_locations - 3; i++) {
+		prod = (cd->mic_locations[i + 1].x - cd->mic_locations[i].x) *
+			(cd->mic_locations[i + 2].y - cd->mic_locations[i].y) -
+			(cd->mic_locations[i + 1].y - cd->mic_locations[i].y) *
+			(cd->mic_locations[i + 2].x - cd->mic_locations[i].x);
+		if (prod)
+			return false;
+	}
+
+	return true;
+}
+
+int tdfb_direction_init(struct tdfb_comp_data *cd, int32_t fs, int channels)
+{
+	struct sof_eq_iir_header_df2t *filt;
+	int64_t *delay;
+	int32_t d_max;
+	int32_t t_max;
+	size_t size;
+	int n;
+	int i;
+
+	/* Select emphasis response per sample rate */
+	switch (fs) {
+	case 16000:
+		filt = (struct sof_eq_iir_header_df2t *)iir_emphasis_16k;
+		break;
+	case 48000:
+		filt = (struct sof_eq_iir_header_df2t *)iir_emphasis_48k;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	/* Allocate delay lines for IIR filters and initialize them */
+	size = channels * iir_delay_size_df2t(filt);
+	delay = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, size);
+	if (!delay)
+		return -EINVAL;
+
+	cd->direction.df2t_delay = delay;
+	for (i = 0; i < channels; i++) {
+		iir_init_coef_df2t(&cd->direction.emphasis[i], filt);
+		iir_init_delay_df2t(&cd->direction.emphasis[i], &delay);
+	}
+
+	/* Unit delay length in Q1.31 seconds */
+	cd->direction.unit_delay = ((int64_t)1 << 31) / fs;
+
+	/* Get max possible mic-mic distance, then from max distance to max time t = d_max / v,
+	 * t is Q1.15, d is Q4.12, v is Q9.0
+	 */
+	d_max = max_mic_distance(cd);
+	t_max = (d_max << 3) / SPEED_OF_SOUND;
+
+#ifdef DEBUG
+	fh = fopen("direction_debug.txt", "w");
```

Comment:
should be done with trace.

Reply:
Yep, also need to think how to get testbench based verification of control updates to work.